pkg/azure/aks/machine_store.go

```diff
@@ -0,0 +1,253 @@
+	MachineTypeSku  string
+	OperatingSystem MachineOperatingSystem
+	Priority        MachinePriority
+}
+
+type MachineStore struct {
+	context context.Context
+	logger  *slog.Logger
+
+	azResourceGroupClient *armresources.ResourceGroupsClient
+	azVMSSClient          *armcompute.VirtualMachineScaleSetsClient
+	azVMSSVmClient        *armcompute.VirtualMachineScaleSetVMsClient
+	azAksClient           *armcontainerservice.ManagedClustersClient
+
+	MachineMap     map[string]*VirtualMachineInfo
+	machineMapLock *sync.RWMutex
+}
+
+func NewMachineStore(parentCtx context.Context, parentLogger *slog.Logger, subscriptionId string, credentials *azidentity.DefaultAzureCredential) (*MachineStore, error) {
+	logger := parentLogger.With("subsystem", "machineStore")
+
+	rgClient, err := armresources.NewResourceGroupsClient(subscriptionId, credentials, nil)
+	if err != nil {
+		logger.LogAttrs(parentCtx, slog.LevelError, "unable to create resource group client", slog.String("err", err.Error()))
+		return nil, ErrClientCreationFailure
+	}
+
+	computeClientFactory, err := armcompute.NewClientFactory(subscriptionId, credentials, nil)
+	if err != nil {
+		logger.LogAttrs(parentCtx, slog.LevelError, "unable to create compute client factory", slog.String("err", err.Error()))
+		return nil, ErrClientCreationFailure
+	}
+
+	containerClientFactory, err := armcontainerservice.NewClientFactory(subscriptionId, credentials, nil)
+	if err != nil {
+		logger.LogAttrs(parentCtx, slog.LevelError, "unable to create container client factory", slog.String("err", err.Error()))
+		return nil, ErrClientCreationFailure
+	}
+
+	ms := &MachineStore{
+		context: parentCtx,
+		logger:  logger,
+
+		azResourceGroupClient: rgClient,
+		azVMSSClient:          computeClientFactory.NewVirtualMachineScaleSetsClient(),
+		azVMSSVmClient:        computeClientFactory.NewVirtualMachineScaleSetVMsClient(),
+		azAksClient:           containerClientFactory.NewManagedClustersClient(),
+
+		MachineMap:     make(map[string]*VirtualMachineInfo),
+		machineMapLock: &sync.RWMutex{},
+	}
+
+	go func() {
+		err := ms.PopulateMachineStore(ms.context)
+		if err != nil {
+			ms.logger.LogAttrs(ms.context, slog.LevelError, "error populating initial machine store", slog.String("error", err.Error()))
+		}
+	}()
+
+	return ms, nil
+}
+
+func (m *MachineStore) getVmInfoByVmName(vmName string) (*VirtualMachineInfo, error) {
+	m.machineMapLock.RLock()
+	defer m.machineMapLock.RUnlock()
+
+	vmInfo := m.MachineMap[vmName]
+	if vmInfo == nil {
+		return nil, ErrMachineNotFound
+	}
+
+	return vmInfo, nil
+}
+
+func (m *MachineStore) getVmInfoFromVmss(ctx context.Context, rgName string, vmssName string, priority MachinePriority, osInfo MachineOperatingSystem) (map[string]*VirtualMachineInfo, error) {
+	vmInfo := make(map[string]*VirtualMachineInfo)
+
+	opts := &armcompute.VirtualMachineScaleSetVMsClientListOptions{
+		Expand: to.StringPtr("instanceView"),
+	}
+	pager := m.azVMSSVmClient.NewListPager(rgName, vmssName, opts)
```

Comment:
Suggestion, to improve the ability to test this method and make it reusable if we need to in the future, pass in the aks client as a parameter to the method and decouple the method from MachineStore.

Reply:
same as above